### ISSUE TITLE
How to transfer pytorch transforms to DALI

### ISSUE THREAD

**Issue by MrNeither** [NVIDIA/DALI#2344]
Hello. I want to transfer video processing from cv and torch to DALI. This option works for me, but it is not optimized
```python
class VideoPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data):
        super(VideoPipe, self).__init__(batch_size, num_threads, device_id, seed=120)
        self.video_reader_resize = ops.VideoReader(
            device="gpu", file_list=data, sequence_length=1, shard_id=0, num_shards=1,
            random_shuffle=True, initial_fill=10, pad_last_batch=True, image_type=types.RGB, dtype=types.FLOAT, 
            file_list_frame_num=False, enable_frame_num=False,enable_timestamps=False)
       
        color_jitter = transforms.ColorJitter(0.8, 0.8, 0.8, 0.2)
        self.transform = transforms.Compose([transforms.ToPILImage(),
                                             transforms.RandomResizedCrop(size=224),
                                             transforms.RandomHorizontalFlip(),
                                             transforms.RandomApply([color_jitter], p=0.5),
                                             transforms.RandomGrayscale(p=0.2),
                                             GaussianBlur(kernel_size=int(21)),  # my class for GaussianBlur in torchvision semantics
                                             transforms.ToTensor()])
    def define_graph(self):
        out = self.video_reader_resize(name="Reader")
        return out

    def init(self):
        self.build()
        self.torch_buffer = torch.empty(self.batch_size, 1, 720, 1280, 3, dtype=torch.float, requires_grad=False)
        self.i = 0
        self.epochs = self.reader_meta()['Reader']['epoch_size']

    def __iter__(self):
        return self

    def __next__(self):
        if self.i <= self.epochs:
            out, labels = self.run()
            self.torch_buffer = feed_ndarray(out.as_cpu().as_tensor(), self.torch_buffer)
            out = torch.stack([self.transform(img) for img in self.torch_buffer.squeeze(1)])
            self.i += 1
            return out, 0
        else:
            raise StopIteration
``` 
I see methods in the ops class,but they are not probabilistic, also I found ops.CoinFlip(), but I don't really understand how to combine this to work as expected.
 To extract a sequence to a frame, I use ops.ElementExtract(device="gpu", element_map=0).
I would greatly appreciate any help!

**Comment by JanuszL** [comment]
Hi,
Unfortunately, not all DALI operators support video data. Among the one you need:
- flip
- Gaussian blur
works.
However RandomResizedCrop and random grayscale (ColorConversion with random saturation value) doesn't support the needed operation. How to approach RandomGrayscale you can find [here](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/hsv_example.html#Using-HSV-to-implement-RandomGrayscale-operator).

In [this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/use_cases/mxnet/mxnet-resnet50.html) you can see how to perform random flip using the CoinFlip operator.

`ElementExtract` may work as we are working on adding convince utility to stack tensors (https://github.com/NVIDIA/DALI/pull/2339). So you would apply `ops.ElementExtract(device="gpu", element_map=0)`, get all frames as a set of separated batches (batch with 0 frames, batch with 1st frame ....), apply all operators on them, and then stack them back.

**Comment by MrNeither** [comment]
Thank you very much for your quick response!
So if I split the batch into elements and perform methods using CoinFlip, will it be a good speed?
And if I have seq_len=1, then ElementExtract will work good, right?

**Comment by JanuszL** [comment]
> So if I split the batch into elements and perform methods using CoinFlip, will it be a good speed?
And if I have seq_len=1, then ElementExtract will work good, right?

It should be good if your `seq_len=1` then you just processing images. I think the `VideoReader` would be the bottleneck as you need to decode random frames from the video - so it means: get to the keyframe, and decode all samples till the end of your desired sequence. If keyframes are not that dense it may take a while.

**Comment by MrNeither** [comment]
I just need to take each frame from the video with a frequency of one per second in the video and process it. I thought shuffle=True would work faster than shuffle=False. Sorry, last question: shuffle=False with seg_len=1 or shuffle=True with seq_len=batch_size will be faster?

**Comment by JanuszL** [comment]
> shuffle=False with seg_len=1 or shuffle=True with seq_len=batch_size will be faster?

I don't think there would be any difference. DALI doesn't reuse any state from the previous sample, so every time it needs to seek to the desired frame and decode it starting from the key frame. We have a plan to optimize the VideoReader for such use case - just decode the whole video and spew all frames but we cannot commit to any schedule when it is available.

**Comment by JanuszL** [comment]
Hi,

`RandomResizedCrop` support now video inputs. I think that concludes the ask.

**Comment by MrNeither** [comment]
Oh, thank you so much!
